changelog.d/10877.feature

```diff
@@ -0,0 +1 @@
+Ensure `(room_id, next_batch_id)` is unique across [MSC2716](https://github.com/matrix-org/matrix-doc/pull/2716) insertion events in rooms to avoid cross-talk/conflicts between batches.
```

Comment:
> Do we need to add a unique constraint to the DB tables?

@erikjohnston If possible, but you shied away from it in https://github.com/matrix-org/synapse/pull/10245#discussion_r669641879

> I think we just want to make `event_id` unique? (Note that event IDs are globally unique). 
> 
> I'm not sure whether we can assert that `(room_id, next_chunk_id)` is unique here, I could imagine that one day a remote server gets confused and reuses a chunk ID or something, at which point we probably *don't* want to explode on insertion of the event.
>
> *-- @erikjohnston, https://github.com/matrix-org/synapse/pull/10245#discussion_r669641879*

Reply:
Oh lols. I'm not sure I agree with myself there, but haven't really thought about it. Anyway, doesn't need to block this  PR.